proposals/0149-migrate-snapshot-epoch-stakes.md

```diff
@@ -0,0 +1,159 @@
+the current epoch's stake cache is be updated during rewards distribution with
+newly rewarded stake accounts and possibly newly created stake accounts as well.
+
+## Detailed Design
+
+### Serialized Data Changes
+
+To be more specific, the missing stake information is the `credits_observed`
+field from the `Stake` struct:
+
+```rust
+struct Stake {
+    delegation: Delegation,
+    credits_observed: u64,
+}
+```
+
+The currently used `epoch_stakes` field in the bank snapshot serializes a map of
+epochs to `EpochStakes` structs which have special logic for serializing stake
+state:
+
+```rust
+struct (BankFieldsToDeserialize | BankFieldsToSerialize) {
+    ..
+    epoch_stakes: HashMap<Epoch, EpochStakes>,
+    ..
+}
+
+struct EpochStakes {
+    #[serde(with = "crate::stakes::serde_stakes_enum_compat")]
+    stakes: Arc<StakesEnum>,
+    total_stake: u64,
+    node_id_to_vote_accounts: Arc<NodeIdToVoteAccounts>,
+    epoch_authorized_voters: Arc<EpochAuthorizedVoters>,
+}
+
+enum StakesEnum {
+    Accounts(Stakes<StakeAccount>),
+    Delegations(Stakes<Delegation>),
+}
+```
+
+When serializing `EpochStakes` in snapshots, all `StakesEnum` variants first map
+stake entry values to their `Delegation` value before serialization. The goal of
+this proposal is to migrate to a new epoch stakes field which maps stake entry
+values to their full `Stake` value before serialization so that
+`credits_observed` will be be included in the snapshot and available after
+snapshot deserialization.
+
+The proposed `new_epoch_stakes` bank snapshot field will instead serialize a map
+of epochs to `VersionedEpochStakes` structs which can be updated in the future
+to serialize different information if needed. This field will be appended to the
+end of the serialized bank snapshot:
+
+```rust
+struct (BankFieldsToDeserialize | BankFieldsToSerialize) {
+    ..
+    new_epoch_stakes: HashMap<Epoch, VersionedEpochStakes>,
+}
+
+enum VersionedEpochStakes {
+    Current {
+        stakes: Stakes<Stake>,
+        total_stake: u64,
+        node_id_to_vote_accounts: Arc<NodeIdToVoteAccounts>,
+        epoch_authorized_voters: Arc<EpochAuthorizedVoters>,
+    },
+}
+```
+
+### Snapshot Update Rollout
+
+Handling snapshot format changes is always a delicate operation to coordinate
+given that old software releases will not be able to deserialize snapshots from
+new software releases properly. The rollout will require two phases:
+
+1. Introduce support for deserializing the migrated epoch stakes field 
+2. Enable serializing epoch stakes to the new field and phase out the old field
+
+During the first phase, validator software will be updated to attempt to
+deserialize the new epoch stakes field appended at the end of the bank snapshot.
+If the field doesn't exist, validators will continue using the old deserialized
+field. If the new epoch stakes field is deserialized successfully, the epoch
+stakes entries from this field will be merged with entries from the old field.
```

Comment:
Do we actually need to support merging the two fields? ie. is there a use-case for having both the old field and the new field populated?

Reply:
We typically keep around like 6 epochs worth of epoch stakes entries. After the migration we could have some delegation only entries still which I think we should keep storing in the old field until enough epochs pass.

Reply:
Thanks for adding this info; much clearer now.